### ISSUE TITLE
(Maybe) Add experiment evaluation to blue-green deployments

### ISSUE THREAD

**Issue by dsbos** [argoproj/argo-rollouts#1121]
# Summary

Consider allowing experiments to be used in blue-green deployments.

It seems that Experiments can be used in canary deployments (to check advancing past a certain step), but that they can't be used similarly in blue-green deployments.

# Use Cases

Here's the use case I'm looking at:
- First, I want to deploy the new (preview) version of my service/application with a _non-production_ configuration that sets some flag so that the service runs in test mode (e.g., pulling from a test queue without disrupting the corresponding production queue).
- Then I want to run some tests against that non-production preview version (and then undeploy it).
- If the tests pass, then I want to start rolling out my preview version with its normal production configuration. I _don't_ need to gradually ramp the new version and the old version down.
- Then I want to switch production traffic from the old stable version to what was the preview version and is the new stable version.

It  looks like I can accomplish that by:
- setting up a canary rollout
- starting at 0% preview / 100% stable
- having a canary step with an Experiment whose pods run the preview version with the test-mode configuration and whose analysis(-es) runs tests against that non-production preview version
- jumping up to 100% canary / 0 % stable (since I don't need to ramp up/down gradually)
- promoting/finalizing the rollout

Since I'm not really using the ramping ability of the canary rollout, it seems that it would be clearer and cleaner if I could just use a blue-green rollout and include the Experiment in the pre-promotion part (gating promotion on success of the Experiment's analyses).
  
However, at least from the current documentation, doing that doesn't seem to be possible. 

---
(Ideally, In my specific case (running experimental pods that are almost the same as the preview pods) the experiment could "inherit" the pod configuration (e.g., image tag and some environment variables/etc.) from the Rollout's main pod configuration (for the preview service) and would need to specify only what was _different_ (e.g., some flag environment variable).)

---

**Message from the maintainers**:

Impacted by this bug? Give it a 👍. We prioritize the issues with the most 👍.

**Comment by jessesuen** [comment]
You're correct to point out that the canary strategy is a superset of the capability of the blue-green (when a traffic router is used). How you described the way to accomplish your use case using canary steps is precisely how I would recommend to others.

It's unlikely we'll change blueGreen syntax to support your proposal, given that the canary strategy already can accomplish it, and the terminology is really just semantics at that point
